src/libraries/System.Private.CoreLib/src/System/Threading/Lock.cs

```diff
@@ -0,0 +1,1199 @@
+                Debug.Assert(_recursionCount == 0);
+                _owningThreadId = currentThreadId.Id;
+                return true;
+            }
+
+            return TryEnterSlow(timeoutMs, currentThreadId);
+        }
+
+        /// <summary>
+        /// Exits the lock.
+        /// </summary>
+        /// <remarks>
+        /// If the calling thread holds the lock multiple times, such as recursively, the lock is exited only once. The
+        /// calling thread should ensure that each enter is matched with an exit.
+        /// </remarks>
+        /// <exception cref="SynchronizationLockException">
+        /// The calling thread does not hold the lock.
+        /// </exception>
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        public void Exit()
+        {
+            var owningThreadId = new ThreadId(_owningThreadId);
+            if (!owningThreadId.IsInitialized || owningThreadId.Id != ThreadId.Current_NoInitialize.Id)
+            {
+                ThrowHelper.ThrowSynchronizationLockException_LockExit();
+            }
+
+            ExitImpl();
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal void ExitImpl()
+        {
+            Debug.Assert(new State(this).IsLocked);
+
+            if (_recursionCount == 0)
+            {
+                _owningThreadId = 0;
+
+                State state = State.Unlock(this);
+                if (state.HasAnyWaiters)
+                {
+                    SignalWaiterIfNecessary(state);
+                }
+            }
+            else
+            {
+                _recursionCount--;
+            }
+        }
+
+        private static bool IsAdaptiveSpinEnabled(int minSpinCount) => minSpinCount <= 0;
+        private static int TickCount => (int)TimerQueue.TickCount64;
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private bool TryEnterSlow(int timeoutMs, ThreadId currentThreadId)
+        {
+            Debug.Assert(timeoutMs >= -1);
+
+            if (!currentThreadId.IsInitialized)
+            {
+                // The thread info hasn't been initialized yet for this thread, and the fast path hasn't been tried yet. After
+                // initializing the thread info, try the fast path first.
+                currentThreadId.InitializeForCurrentThread();
+                Debug.Assert(_owningThreadId != currentThreadId.Id);
+                if (State.TryLock(this))
+                {
+                    goto Locked;
+                }
+            }
+            else if (_owningThreadId == currentThreadId.Id)
+            {
+                Debug.Assert(new State(this).IsLocked);
+
+                uint newRecursionCount = _recursionCount + 1;
+                if (newRecursionCount != 0)
+                {
+                    _recursionCount = newRecursionCount;
+                    return true;
+                }
+
+                throw new LockRecursionException(SR.Lock_Enter_LockRecursionException);
```

Comment:
I don't think I'm following how this gets thrown. It would mean _recursionCount was -1. I don't see a test in LockTests for LockRecursionException.

Reply:
Oh - you're using 32 bit overflow [like Native AOT ](https://github.com/dotnet/runtime/issues/85128) .. since this is a new type, how about picking something smaller that is still always going to mean a bug in the app,  but small enough that it won't cause a hang in the app (and also becomes feasibly unit testable). Eg., 65K

Reply:
in fact that's what the comment suggests "is expected to be high enough that it would typically not be reached when the lock is used properly."